admin/server/github.go

```diff
@@ -88,53 +134,254 @@ func (s *Server) githubConnect(w http.ResponseWriter, r *http.Request, pathParam
+		return
+	}
+
+	// install/update setupAction
+
+	code := qry.Get("code")
+	if code == "" {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
+
+	// exchange code to get an auth token and create a github client with user auth
+	githubClient, err := s.userAuthGithubClient(ctx, code)
+	if err != nil {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
+
+	// verify there is no spoofing and the user is a collaborator to the repo
+	// There may not be a need for this because we are not storing any state wrt installation
+	// Keeping this for future proofness if we start storing some states
+	gitUser, isCollaborator, err := s.isCollaborator(ctx, account, repo, githubClient)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("failed to track github install: %s", err), http.StatusInternalServerError)
+		http.Error(w, fmt.Sprintf("failed to verify ownership: %s", err), http.StatusUnauthorized)
 		return
 	}
 
+	if !isCollaborator {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
+
+	// save github user name
+	if claims.OwnerID() != "" {
+		if user, err := s.admin.DB.FindUser(ctx, claims.OwnerID()); err == nil { // ignoring error since can be called by user not logged into rill cloud
+			_, err := s.admin.DB.UpdateUser(ctx, user.ID, user.DisplayName, user.PhotoURL, gitUser.GetLogin())
+			if err != nil {
+				s.logger.Error("failed to update user's github username")
+			}
+		}
+	}
+
+	// Verify that user installed the app on the right repo and we have access now
+	installationID, err := s.admin.GetGithubInstallation(ctx, remoteURL)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to check github repo status: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	if installationID == 0 {
+		// no access
+		// Redirect to UI retry page
+		redirectURL, err := url.Parse(s.urls.githubConnectRetry)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create retry request url: %s", err.Error()), http.StatusInternalServerError)
+			return
+		}
+
+		qry := redirectURL.Query()
+		qry.Set("remote", remoteURL)
+		redirectURL.RawQuery = qry.Encode()
+		http.Redirect(w, r, redirectURL.String(), http.StatusTemporaryRedirect)
+	}
+
 	// Redirect to UI success page
-	redirectURL, err := url.JoinPath(s.opts.FrontendURL, "/-/github/connect/success")
+	http.Redirect(w, r, s.urls.githubConnectSuccess, http.StatusTemporaryRedirect)
+}
+
+// githubAuthLogin starts user authorisation of github app.
+// In case github app is installed by another user, other users of the repo need to separately authorise github app
+// where this flow comes into picture.
+// Some implementation details are copied from auth package.
+// It's implemented as a non-gRPC endpoint mounted directly on /github/auth/login.
+func (s *Server) githubAuthLogin(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	// Generate random state for CSRF
```

Comment:
It should not start this flow if the user is not logged into Rill. Right now it only checks `claims.OwnerType() != auth.OwnerTypeUser` in the callback.

I guess if the user is not logged into Rill in the browser, it should redirect to Rill login, and after login, redirect back to this? I think the auth endpoints already support passing a `redirect` option.

Reply:
Yeah makes sense. Although a user is expected to be logged into Rill when they are redirected to this page. 
We also need to make a similar change in `github/connect` as well. This is especially required since we removed redirecting from a webpage which already had this option.

Reply:
I don't think the assumption is safe. They might have logged out of Rill (or changed default browsers, or cleared browser state) since authenticating the CLI (which they may have done months earlier). So best to redirect for auth everywhere that we rely on it.